clang/lib/Sema/SemaOverload.cpp

```diff
@@ -1595,8 +1606,21 @@ static bool IsOverloadOrOverrideImpl(Sema &SemaRef, FunctionDecl *New,
 
         // Allow overloading of functions with same signature and different CUDA
         // target attributes.
-        if (NewTarget != OldTarget)
+        if (NewTarget != OldTarget) {
+          // Special case: non-constexpr function is allowed to override
+          // constexpr virtual function
```

Comment:
Would it make sense to make this relaxation a bit more generic and consider functions to be the same if NewTarget/OldTarget are the same if we ignore implicit attributes?

Reply:
That seems to be too generic. The implicit host/device attributes are not just added to constexpr functions. They are also added to inherited ctors, e.g. https://godbolt.org/z/PvGfzj5ej . An inherited ctor may have only one implicit device attribute, which should be treated as a device function, like an explicit device attribute. I tends to think the special handling needed by constexpr pure-virtual function is a rare exception.

Reply:
I see. So, we have a distinction between implicit attribute derived from the explicit one, and the implicit attributes conjured by compiler for the compiler-generated code like special functions and lambdas. AFAICT we have no good way to do so. E.g. telling apart inherited constructor with explicit HD attributes and a default constructor with implicit HD attributes would require more digging.

OK. Let's keep it narrow for now. We can consider relaxation if/when we have actual use cases for it.